### PR TITLE
Cluster info metrics

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Lock contention on gossip locks are causing lots of latency.

#### Summary of Changes

Add metrics about which users are taking lock time.

Fixes #

### CODE REVIEW

**Comment by garious** [comment]
Saving the guard in a variable will cause the lock to be held until that variable falls out of scope. Easy to imagine unintended consequences here, such as deadlocks or additional contention.

**Comment by sakridge** [comment]
> Saving the guard in a variable will cause the lock to be held until that variable falls out of scope. Easy to imagine unintended consequences here, such as deadlocks or additional contention.

Well yes, I believe I've preserved the same critical section as the original code. Is there a better way?

**Comment by garious** [comment]
Passing a closure to gossip_w would allow you to preserve the original semantics without requiring the caller to add an explicit call to `drop`.  Also, are you planning to keep this code around for the long haul or just debugging and then revert? If long-term, can you think a more descriptive function name? `gossip_w` was a bad variable name. As a function name, it's so much worse.

**Comment by sakridge** [comment]
> Passing a closure to gossip_w would allow you to preserve the original semantics without requiring the caller to add an explicit call to `drop`. Also, are you planning to keep this code around for the long haul or just debugging and then revert? If long-term, can you think a more descriptive function name? `gossip_w` was a bad variable name. As a function name, it's so much worse.

Yea, I thought of that, closure likely has worse performance though. Names are easily changed.

**Comment by garious** [comment]
I can't think of a reason a closure would be slower, theoretically. It's stack-allovated. Maybe the compiler implementation prevents inlining or using the wrong one forces an implicit copy, but that's really stretching for something to worry about. It's nothing compared to how this current use of carefully placed `drop`s and confusing function names that'd make newcomers afraid to touch this code.

File: core/src/cluster_info.rs [hunk 1]
```diff
@@ -108,6 +108,53 @@ pub struct DataBudget {
                   // used to detect when to up the bytes budget again
 }
 
+#[derive(Default)]
+struct Counter(AtomicU64);
+
+impl Counter {
+    fn add_m(&self, x: &mut Measure) {
```
Comment (garious) [inline core/src/cluster_info.rs:115]: `_m` means what here? Mut? Measure?

File: core/src/cluster_info.rs [hunk 2]
```diff
@@ -108,6 +108,53 @@ pub struct DataBudget {
...
+        x.stop();
+        self.0.fetch_add(x.as_us(), Ordering::Relaxed);
+    }
+    fn add_r(&self, x: u64) {
```
Comment (garious) [inline core/src/cluster_info.rs:119]: `_r` means what here? Callable from read-only? Relaxed ordering? Can you give these better names or add code comments?

**Comment by sakridge** [comment]
> 
> 
> I can't think of a reason a closure would be slower, theoretically. It's stack-allovated. Maybe the compiler implementation prevents inlining or using the wrong one forces an implicit copy, but that's really stretching for something to worry about. It's nothing compared to how this current use of carefully placed `drop`s and confusing function names that'd make newcomers afraid to touch this code.

Yea, mostly inlining. I have a better solution. Simple struct that contains the lock and also handles the timer.

**Pull request merged**